### PR TITLE
Structured references

### PR DESCRIPTION
**Pull request by rmdort**
Implemented some of the feature of structured reference

https://support.microsoft.com/en-us/office/using-structured-references-with-excel-tables-f5ed2452-2337-4f71-bed3-c8ae6d2b276e

1. Table, Column detection
2. This row `@` detection

Would be great to get some feedback. 

Todo

1. Detect `#Total` and `#Data`, `#ThisRow`
2. Range formulas `DeptSalesFYSummary[[2014]:[2012]]`

### CODE REVIEW

**Review by daniellacosse — APPROVED** [review]

**Comment by LesterLyu** [comment]
Will take a look in this weekend.

**Review by LesterLyu — COMMENTED** [review]
## There are more things need to be done:

### Pound sign support `#`

e.g. `DeptSales[[#Totals],[Sales Amount]]`, `DeptSales[[#Headers],[Sales Amount]]`, `DeptSales[#All]`, `DeptSales[#This Row]`
This special item specifier: | Refers to:
-- | --
`#All` | The entire table, including column headers, data, and totals (if any).
`#Data` | Just the data rows.
`#Headers` | Just the header row.
`#Totals` | Just the total row. If none exists, then it returns null.
`#This Row` or `@` or`@[Column Name]` | Just the cells in the same row as the formula. These specifiers can’t be combined with any other special item specifiers. Use them to force implicit intersection behavior for the reference or to override implicit intersection behavior and refer to single values from a column.Excel automatically changes #This Row specifiers to the shorter @ specifier in tables that have more than one row of data. But if your table has only one row, Excel doesn’t replace the #This Row specifier, which may cause unexpected calculation results when you add more rows. To avoid calculation problems, make sure you enter multiple rows in your table before you enter any structured reference formulas.

### Column name with special character support
- Space ` ` support
- Use an escape character: `[`, `]`, `#`, `'` will be `'[`, `']`, `'#`, `''` in formulas
- More?

### At `@`
Support single `@`: `SUM(DeptSales[@])` (mentioned in *Pound sign support*)

### Ranges
DeptSalesFYSummary[[2014]:[2012]]

File: grammar/parsing.js [hunk 1]
```diff
@@ -376,9 +383,40 @@ class Parsing extends EmbeddedActionsParser {
...
+            $.CONSUME(OpenSquareParen);            
+            $.MANY(() => {
+                thisRow = !!$.CONSUME(At).image
+            })
+            let isNested = false
+            $.OPTION2(() => {
+                isNested = !!$.CONSUME(OpenSquareParen).image;
+            })
+            $.AT_LEAST_ONE(() => {
+                columnName = $.CONSUME(Name).image;
```
[the remark this answers is not in the document]
Comment (rmdort) [reply grammar/parsing.js:405]: Could you give some advice on how to add support for space token. I have tried doing

````
const StructuredRefName  = createToken({
    name: 'StructuredRefName',
    pattern: /[a-zA-Z_\s][a-zA-Z0-9_.\s?]*/,
});
````

But unable to parse it properly.

**Comment by rmdort** [comment]
Closing this.  Will issue a new PR

File: grammar/parsing.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (LesterLyu) [reply grammar/parsing.js:405]: You will want to combine the token with `[]`, i.e. `/\[.....\]/`

**Pull request closed without merging**